hwy/targets.cc

```diff
@@ -292,24 +299,25 @@ int64_t DetectTargets() {
     if ((flags & kGroupSSSE3) == kGroupSSSE3) {
       bits |= HWY_SSSE3;
     }
+#if HWY_ARCH_X86_32
+    if ((flags & kGroupSSE2) == kGroupSSE2) {
+      bits |= HWY_SSE2;
+    }
+#endif
   }
 
-  // Clear bits if the OS does not support XSAVE - otherwise, registers
-  // are not preserved across context switches.
+  // Clear AVX2/AVX3 bits if the OS does not support XSAVE - otherwise,
+  // registers are not preserved across context switches.
   if (has_osxsave) {
     const uint32_t xcr0 = ReadXCR0();
     const int64_t min_avx3 = HWY_AVX3 | HWY_AVX3_DL;
     const int64_t min_avx2 = HWY_AVX2 | min_avx3;
-    // XMM
-    if (!IsBitSet(xcr0, 1)) {
-      bits &= ~(HWY_SSSE3 | HWY_SSE4 | min_avx2);
-    }
-    // YMM
-    if (!IsBitSet(xcr0, 2)) {
+    // XMM, YMM
+    if((xcr0 & 0x06) != 0x06) {
```

Comment:
Shouldn't we still be disabling SSE2/SSSE3/SSE4 if the XMM regs aren't being saved by the OS?

Reply:
The lower 128 bits of the XMM registers are guaranteed to be saved on x86_64, even if bit 1 of XCR0 is not set, as the operating system will save the lower 128 bits of XMM0-XMM15 on x86_64 through another mechanism if bit 1 of XCR0 is not set.

The AMD64 ABI also guarantees that at least the lower 128 bits the XMM0-XMM15 registers and the SSE2 instructions are always available to programs, regardless of the value of the XCR0 register.

Reply:
I see, was not aware of that, thanks for adding the comment.